spring-integration-core/src/main/java/org/springframework/integration/channel/FluxMessageChannel.java

```diff
@@ -70,11 +72,12 @@ protected boolean doSend(Message<?> message, long timeout) {
 	public void subscribe(Subscriber<? super Message<?>> subscriber) {
 		this.subscribers.add(subscriber);
 
-		this.flux.doOnCancel(() -> this.subscribers.remove(subscriber))
+		this.flux.doFinally((signal) -> this.subscribers.remove(subscriber))
 				.retry()
 				.subscribe(subscriber);
 
-		this.publishers.values().forEach(ConnectableFlux::connect);
+		this.publishers.values()
```

Comment:
could you please explain the use of `ConnectableFlux` here?

Reply:
I need `ConnectableFlux` to subscribe to provideded publishers in lazy-manner: if this `FluxMessageChannel` is subscribed already, I do `connect()` - otherwise I `connect()` to all the provided publishers in its `subscribe()`.
My point is to be able to propagate events from upstream publishers into subscribers of this `FluxMessageChannel` honoring an internal logic of the `MessageChannel` implementation.

I couldn't find any other way in Reactor how to achieve a lazy subscription task unless this `ConnectableFlux`.

And with that it looks like I can't use `Disposables.composite()` since I don't have a way to remove one from that composite when such an upstream published is finished.

I'm OK to change the logic if you can help with that.

Thanks

P.S. Pushed `flux.onComplete()` BTW and removed reference to subscribers since it is not necessary any more.
👍